src/workerd/util/sqlite.c++

```diff
@@ -438,16 +447,62 @@ kj::Own<sqlite3_stmt> SqliteDatabase::prepareSql(
           } else {
             SQLITE_CALL_FAILED("sqlite3_step()", err);
           }
-
-          // Reduce `sqlCode` to include only what we haven't already executed.
-          sqlCode = kj::StringPtr(tail, sqlCode.end());
-          continue;
         }
-        break;
+
+        // Now set statementResult as the "result" to be either returned or step()-ed through
+        result = statementResult;
+      }
+
+      // Reduce `sqlCode` to include only what we haven't already executed.
+      sqlCode = kj::StringPtr(tail, sqlCode.end());
+    }
+
+    // After processing the entire input, throw if we never got a single result
+    SQLITE_REQUIRE(result != nullptr, "SQL code did not contain a complete statement.", sqlCode);
+    return ownSqlite(result);
+  }
+}
+
+int SqliteDatabase::ingestSql(Regulator& regulator, kj::StringPtr sqlCode) {
+  // Set up the regulator that will be used for authorizer callbacks while preparing this
+  // statement.
+  KJ_ASSERT(currentRegulator == nullptr, "recursive prepareSql()?");
+  KJ_DEFER(currentRegulator = nullptr);
+  currentRegulator = regulator;
+  const char* start = sqlCode.begin();
+  const char* tail = start;
+
+  SQLITE_REQUIRE(sqlite3_get_autocommit(db) == 0, "sql.ingest must be run inside a transaction");
+
+  // While there are still valid
+  while (tail != sqlCode.end() && sqlite3_complete_inner(tail, 1)) {
+    // Compile the next statement in the SQL
+    sqlite3_stmt* result = nullptr;
+    SQLITE_CALL(sqlite3_prepare_v3(db, sqlCode.begin(), sqlCode.size(), 0, &result, &tail));
+
+    // If it is a real statement
+    if (result != nullptr) {
+      // Fully step() through what's in 'result' before returning/processing the new statement
+
+      SQLITE_REQUIRE(sqlite3_bind_parameter_count(result) == 0,
+                     "sql.ingest does not support parameter binding");
+
+      // A single step is sufficient to execute the query, since we don't care about its output
+      int err = sqlite3_step(result);
+      if (err == SQLITE_DONE) {
+        // good
+      } else if (err == SQLITE_ROW) {
+        // Intermediate statement returned results. We will discard.
+      } else {
+        SQLITE_CALL_FAILED("sqlite3_step()", err);
+      }
     }
 
-    return ownResult;
+    sqlCode = kj::StringPtr(tail, sqlCode.end());
   }
+
+  // Return how many characters were processed
```

Comment:
Unfortunately, you are returning a number of UTF-8 bytes, but JavaScript will of course need a measure of UTF-16 code units in order to compute a substring.

Perhaps the function should actually return the leftover text, to avoid ambiguity?

Speaking of, when streaming in SQL, how will you deal with multi-byte characters being split across a buffer boundary?

Reply:
> Perhaps the function should actually return the leftover text, to avoid ambiguity?

Implemented this

> Speaking of, when streaming in SQL, how will you deal with multi-byte characters being split across a buffer boundary?

Would appreciate some suggestions here, I've pushed a commit with a failing test case:

<img width="593" alt="image" src="https://github.com/cloudflare/workerd/assets/23264/04715cf8-8240-4ba8-9da5-2286be5983f6">

Maybe instead of `substring` I should be converting to an array buffer and concatenating that way?

Reply:
Hmm, I'd expect your use of `TextDecoder` in streaming mode to solve this... presumably it should buffer any incomplete UTF-8 sequences?

Reply:
Yes it turns out it was me doing `.substring` before calling `TextEncoder.encode` to send the binary. So I was mangling it before encoding it.

Reply:
Yes it turns out it was me doing `.substring` before calling `TextEncoder.encode` to send the binary. So I was mangling it before encoding it.